Create and activate product attributes lookup table in data migration

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Until now the product attributes lookup table had to be created and its usage had to be activated manually, via the tools page. This pull request does the following changes:

1. Remove the tools to create and delete the table (but keep the tool to regenerate the table data for one or for all the products).

2. Add a data migration that triggers the table creation and filling (unless the table already existed, in this case nothing happens).

3. After the migration finishes, activate the table usage for product filtering (site admin can still disable it via _Settings - Products - Advanced_); again, if the table already existed, nothing happens (the setting is unchanged). The setting will be left disabled if the regeneration is abored manually (see below).

Additionally, the following has been added in order to help in the migration process:

1. A new filter, `woocommerce_attribute_lookup_regeneration_step_size`, that allows to alter the batch size for the process (the number of products that will be processed each time that Action Scheduler runs one migration step). This filter is run on every step, thus it can be used to alter the batch size even if the regeneration process has been already started.

2. If the regeneration process is aborted manually (by deleting the `woocommerce_attribute_lookup_regeneration_in_progress` option), the _Products - Advanced_ settings section will still allow to enable the lookup table usage, but will display a warning discouraging to do so because the lookup table will be in an inconsistent state. The warning is shown when the `woocommerce_attribute_lookup_regeneration_aborted` option exists, and the option is deleted when the table regeneration is triggered again via the tools page.

![image](https://user-images.githubusercontent.com/937723/149765098-830d9792-e6b1-46e0-b9fa-6158a934c557.png)


Closes #30564.
Closes #30565.

### How to test the changes in this Pull Request:

1. If you had tested the product attributes lookup table mechanism already and the table still exists in your db, delete it from _Status - Tools - Delete the product attributes lookup table_ (you need to do this from `trunk`).

2. Switch to the branch of this pull request and go to _Status - Tools_. You shouldn't see any entry related to the product attributes lookup table.

3. Trigger the WooCommerce db update from the update available warning in the admin area.

3. After the update finishes you'll see that the old tools page entry "Regenerate the product attributes lookup table" is back. If you have a lot of products in your db the "Regenerate" button will be disabled and will show a counter of processed products, this means that the table regeneration is still in progress.

4. Once the table filling is complete the "Regenerate" product should appear as enabled. After that has happened go to _Settings - Products - Advanced_ and verify that "Enable table usage" is set.

5. Verify that the feature still works as expected (enabling and disabling the table usage via settings, regenerating the table from tools page for one product or for all products).

Now, to test the case of a db update when the table already exists:

6. Disable the table usage via settings.

7. Run the following in command line: `wp option set woocommerce_db_version 6.0.0`

8. Trigger the db update from the banner again, this time nothing related to the lookup table happens (the "Regenerate" tool has its button enabled right after the db update finishes).

9. Verify that the table usage is still disabled in settings.

To test the new filter:

10. Add the following somewhere (e.g. at the end of `woocommerce.php`):

```php
add_filter('woocommerce_attribute_lookup_regeneration_step_size', function() {
	return 5;
});
```

Verify that the regeneration happens now in increments of five products (hint: use `wp option get woocommerce_attribute_lookup_processed_count` to easily get the count of products already processed). If you change the value mid-regeneration you should see that the new value is used from the following regeneration step.

To test the new warning message:

11. Abort an already started regeneration by running `wp option delete woocommerce_attribute_lookup_regeneration_in_progress`.

12. Verify that the _Products - Advanced_ settings page now shows the warning (you may need to reload a couple of times to give Action Scheduler a chance to run one last time).

13. Start a new regeneration process, let it finish, and verify that the warning has disappeared.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Add - Data migration to create and activate the product attributes lookup table.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Create and activate product attributes lookup table in data migration

Until now the product attributes lookup table had to be created
and its usage had to be activated manually, via the tools page.
This commit does the following:

1. Remove the tools to create and delete the table
   (but keeps the tool to regenerate the table data for one or for
   all the products)

2. Creates a data migration that triggers the table creation and
   filling (unless the table already existed) for Woo 6.1

3. After the migration finishes, activates the table usage for
   product filtering (site admin can still disable it via
   Settings - Products - Advanced)

Konamiman committed:
Sort arrays before comparison in unit test

Without this a test fails in GitHub
("Run unit tests with code coverage" task).

barryhughes commented:
✍🏽 Hold on merging p1637741479224500-slack-C0E1AV8T0 (internal ref).

barryhughes commented:
🔬 **Observation regarding `OR` filtering**

There seems to be a quirk when a filter is configured to use the `OR` query type. Setup:

- I created an attribute called _'complexity'._
- Several terms belonging to this attribute are applied to lots of different products.
- One term called _'custom'_ is applied to just a single product.

With lookup table usage disabled and if I select _'custom'_ from this filter, then the widget looks like this (matching my expectations):

<img width="291" alt="01-no-table-usage" src="https://user-images.githubusercontent.com/3594411/143311450-4440a9bb-064c-4255-8c38-f9db828c9863.png">

With lookup table usage _enabled,_ the widget will look like this:

<img width="306" alt="02-with-table-usage" src="https://user-images.githubusercontent.com/3594411/143311625-5ae29a1a-cb8c-4ad9-83dc-94389079f953.png">

There are two problems here:

- Apart from the selected term, the only other terms I now see are those that happen to be applied to the same product(s) as the selected term. Since we are using `OR` filtering, though, I'd expect _all_ attribute terms to be listed (except perhaps for those that are not applied to any products).
- The counts are incorrect. For example, if I were to additionally select _'fiendish'_ I would in fact see 215 products (and if I go ahead and do that, the counts are corrected).

barryhughes commented:
🔬 **Observation: regenerating lookup data for a single product**

Very minor but noting for completeness: if I use the _regenerate the product attributes lookup table_ tool **for a single product** then (as I understand it) the relevant entries are generated by the time the page reloads, but we still see a message reading:

> Product attributes lookup table data is regenerating

In this particular case, it might be better for it to read:

> Product attribute data for <Product XYZ> was successfully regenerated.

barryhughes commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -169,6 +169,10 @@ class WC_Install {
 			'wc_update_600_migrate_rate_limit_options',
 			'wc_update_600_db_version',
 		),
+		'6.1.0' => array(
```
> Scheduled for WooCommerce 6.2

Given this note, should we adjust to `'6.2.0'` (and if so also adjust `wc_update_610_db_version()`)?

Konamiman replied:
Yep, makes sense. Fixed.

Konamiman replied:
Yep, makes sense. Fixed.

barryhughes reviewed (commented):
Left a couple of notes to check on, but looking good 👍🏽

Konamiman committed:
Fix: bad counters in the filter by attribute widget for OR.

When a filter by attribute widget was configured as OR, the query
to return the product counts was overfiltering and returning
less results than it should. It now works as in the case of not
using the lookup table.

Konamiman committed:
Temporary fix for a unit test.

The "Advanced" section in the product settings page exists only
if the product attributes lookup table exists. The table will
eventually exist for everybody, so a temporary check has been added.

Konamiman committed:
Change db update version from 6.1 to 6.2.

Konamiman commented:
> Very minor but noting for completeness: if I use the regenerate the product attributes lookup table tool for a single product then (as I understand it) the relevant entries are generated by the time the page reloads, but we still see a message reading:

Be careful, reloading the tools page is tricky. If you just reload, the page URL will still have the `action` query string parameter set, this you are effectively triggering the regeneration process again. You need to manually remove the parameter from the URL before reloading.

Konamiman commented:
> There seems to be a quirk when a filter is configured to use the OR query type

Indeed, looks like it was "overfiltering" in this case. I think I have fixed it.

barryhughes reviewed (dismissed):
LGTM!

- Not merging (per instructions)
- Checks are failing, but as far as I can determine the fails are unrelated

hossamadel99 commented:
Hello ,

Followed all steps but when trying to filter any attribute it takes about 20 seconds, the page refreshes in two times, one instantly without any filtration but after 20 seconds succeed.

Any guide please regarding the speed with ( attributes lookup table )

Konamiman commented:
Hi @hossamadel99 , are you ensuring that the table regeneration has finished before attempting to actually load the products page?

hossamadel99 commented:
> Hi @hossamadel99 , are you ensuring that the table regeneration has finished before attempting to actually load the products page?

Yes & got (Regenerate) button

Konamiman commented:
@hossamadel99 I'd need more data in order to troubleshoot this. Is there any testing site I can take a look at? Or alternatively, could you please send me a product dump? You can reach me at nestor.soriano@automattic.com

Konamiman commented:
Just noting here for reference that @hossamadel99 did indeed contact me via email and confirmed me that after upgrading to WooCommerce 6.0 the performance issue disappeared. Thanks!

Konamiman committed:
Add a table regeneration step size filter and a "regeneration aborted" warning

Add a new filter: woocommerce_attribute_lookup_regeneration_step_size,
it allows to choose how many products will be processed on each
step when the product attributes lookup table is being regenerated,
the default is the previous fixed value, 10.

Also when the regeneration is manually aborted (via deleting the
woocommerce_attribute_lookup_regeneration_in_progress option), the
products - advanced settings section will still allow to enable the
usage of the table, but a warning will be shown because the table
will be in an inconsistent state.

Konamiman committed:
Change db update version from 6.2 to 6.3.

Also rename the db update function to include the Woo version number.

jeffstieler commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2335,3 +2337,28 @@ function wc_update_600_migrate_rate_limit_options() {
+ *
+ * @return false Always false, since the LookupDataStore class handles all the data filling process.
+ */
+function wc_create_product_attributes_lookup_table() {
```
Should we continue the pattern of including the version number in the update function name here?

jeffstieler commented on plugins/woocommerce/woocommerce.php:
```diff
@@ -60,3 +60,7 @@ function wc_get_container() : \Psr\Container\ContainerInterface {
+
+/*add_filter('woocommerce_attribute_lookup_regeneration_step_size', function() {
+	return 5;
+});*/
```
Looks like this might have been included by mistake.

jeffstieler reviewed (commented):
Saw some minor things.

barryhughes commented:
A little unsure if this is a problem or not (or even if it has to be addressed right away, if we do see it as a problem). At any rate, I wanted to note that if I use this tool:

<img width="809" alt="verify-base-database-tables" src="https://user-images.githubusercontent.com/3594411/150027588-4503018a-31e3-4ea9-b0b8-209790245e1d.png">

Then it reports success:

<img width="813" alt="verify-base-database-tables-successful" src="https://user-images.githubusercontent.com/3594411/150027670-8e5d27d8-bb48-404a-9537-6961ca17995d.png">

However, the product attributes lookup table is not present/has not yet been created. *Should* it be covered by this tool, or will it remain a distinct thing?

barryhughes commented on plugins/woocommerce/src/Internal/ProductAttributesLookup/LookupDataStore.php:
```diff
@@ -648,4 +657,28 @@ public function set_regeneration_in_progress_flag() {
+	 *
+	 * @return bool True if the last lookup table regeneration process was aborted.
+	 */
+	public function regeneration_was_aborted() {
```
Minor thing, but we could use return types here:

```suggestion
	public function regeneration_was_aborted(): bool {
```

barryhughes reviewed (dismissed):
LGTM. 

I left a couple of notes, so am not merging so you can review those first of all; but if no action is needed and we're otherwise ready to merge please go ahead.

Konamiman committed:
Enable the product attributes lookup table after data regeneration

After a regeneration of the product attributes lookup table finishes
successfully the table usage will be automatically enabled, except
if the regeneration was aborted via the deletion of the
woocommerce_attribute_lookup_regeneration_in_progress option.

Konamiman committed:
Set the return type of LookupDataStore::regeneration_was_aborted to bool

Konamiman commented:
@barryhughes The attributes lookup table should probably be included in the list of core WooCommerce tables, indeed; but that's also the case for a bunch of other already existing WooCommerce tables. So fixing this would better be done in a separate project and pull request.

barryhughes reviewed (approved):
Looks good

barryhughes commented:
Going ahead with merge (just noting this because the description still states it shouldn't be merged, but per recent-ish team discussion this work is good to go).
